contracts/TaskBook.sol

```diff
@@ -5,83 +5,158 @@ import './AccountManager.sol';
+		t.state = newState;
+		TaskStateChange(taskID, newState);
+		return true;
+	}
+
+	//Solver registers for tasks, if first to register than automatically selected solver
+	//0->1
 	function registerForTask(uint taskID, uint minDeposit, bytes32 randomBitsHash) returns(bool) {
 		require(balances[msg.sender] >= minDeposit);
-		Task t = tasks[taskID];
+		Task storage t = tasks[taskID];
 		require(!(t.owner == 0x0));
-		require(t.numSolvers < maxSolvers);
+		require(t.state == 0);
+		require(t.selectedSolver == 0x0);
+		t.selectedSolver = msg.sender;
 		solverRandomBitsHash[msg.sender][taskID] = randomBitsHash;
-		//random = sha3(random, block.blockhash(block.number-1));
-		t.solvers[t.numSolvers] = msg.sender;
-		t.numSolvers++;
-		log0(bytes32(sha3(msg.sender)));
-		return true;
-	}
-
-	//Task Issuer tells TrueBitExchange to select a solver
-	function selectSolver(uint taskID) returns (bool) {
-		Task t = tasks[taskID];
-		require(msg.sender == t.owner);
-		//uint random = uint(block.blockhash(block.number-1));
-		//address randomSolver = t.solvers[uint(random) % t.numSolvers];
-		address solver = t.solvers[0];
-		t.selectedSolver = solver;
-		SolverSelected(taskID, solver, t.taskData, t.minDeposit);
+		t.state = 1;
+		log0(randomBitsHash);
+		SolverSelected(taskID, msg.sender, t.taskData, t.minDeposit);
 		return true;
 	}
 
 	//Selected solver submits a solution to the exchange
-	function commitSolution(uint taskID, bytes32 solutionHash) returns (bool) {
-		Task t = tasks[taskID];
+	//1->2
+	function commitSolution(uint taskID, bytes32 solutionHash0, bytes32 solutionHash1) returns (bool) {
+		Task storage t = tasks[taskID];
 		require(t.selectedSolver == msg.sender);
-		solutions[msg.sender][taskID] = solutionHash;
-		SolutionCommitted(taskID, t.minDeposit, t.taskData, msg.sender);
+		require(t.state == 1);
+		Solution storage s;
+		s.solutionHash0 = solutionHash0;
+		s.solutionHash1 = solutionHash1;
+		solutions[taskID] = s;
+		t.state = 2;
+		SolutionsCommitted(taskID, t.minDeposit, t.taskData, msg.sender);
 		return true;
 	}
 
 	//Verifier submits a challenge to the solution provided for a task
-	function commitChallenge(uint taskID, uint minDeposit, address solver, bytes32 intentHash) returns (bool) {
+	//Verifiers can call this until task giver changes state or timeout
+	function commitChallenge(uint taskID, uint minDeposit, bytes32 intentHash) returns (bool) {
 		require(balances[msg.sender] >= minDeposit);
-		require(!(solutions[solver][taskID] == 0x0));
-		Task t = tasks[taskID];
-		require(t.numChallengers < maxChallengers);
-		verifierIntent[msg.sender][taskID] = intentHash;
-		t.challengers.push(msg.sender);
-		log0(sha3(msg.sender));
+		Task storage t = tasks[taskID];
+		require(t.state == 2);
+		t.challenges[msg.sender] = intentHash;
 		return true;
 	}
-}
\ No newline at end of file
+
+	//Verifiers can call this until task giver changes state or timeout
+	function revealIntent(uint taskID, uint intent) returns (bool) {
```

Comment:
Challengers would need to also supply what *they* think is the correct solution. Otherwise, there is nothing to play a verification with.

Reply:
Yes when they commit to a challenge they commit a hash of their intent, the intent reveals which one they think is the wrong one, therefore, the assumption they think the other is correct.

Reply:
What about in the case of unforced-errors? 

We need to allow challengers to provide their own solutionHash in case they disagree with both of the Solver’s solutions (which would be the case in an unforced-error).

Reply:
Wouldn't the verifier then end up recomputing that hash anyways when participating in the verification game? Providing the solution hash would be the end of the computation, and I don't think it works backwards.

I agree that the hash *has* to get computed at some point. But I think it can be deferred till later like in the verification game function itself.

Reply:
I think the Truebit paper had challengers include their solution hash. But what you’re saying makes sense. Think we can leave it until the verification game.